### PR TITLE
ref #826: Improve table editor's save functionality

### PR DESCRIPTION
**Pull request by pixeljunkie**
| Q             | A
| ------------- | ---
| Branch        | 7.1.x
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed issues  | chameleon-system/chameleon-system#826
| License       | MIT

### Summary by CodeRabbit

**New Feature:**
- Introduced a new custom event `tableEditorBeforeSaveEvent` in `tableEditor.js`. This event is dispatched before saving operations, allowing other parts of the code to perform necessary actions.
- Added event listeners for `tableEditorBeforeSaveEvent` in `editor.html.twig`. These listeners retrieve data from CKEDITOR instances and set it as the value of specific elements.

> "Hoppity hop, with a skip and a jump,
> Our code takes a leap, not a single bump. 🐇
> With events so neat, and listeners in place,
> We're saving our data with style and grace. 🎉
> So here's to the changes, may they run smooth and fast,
> In the world of coding, we're having a blast! 💻🚀"

### CODE REVIEW

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 1]
```diff
@@ -17,7 +17,11 @@ var _trans_icon = '';
 var _trans_link = '';
 var _trans_password = '';
 var _trans_drag = '';
-
+let tableEditorBeforeSaveEvent = new CustomEvent('tableEditorBeforeSaveEvent', {
```
Comment (finn-matti) [inline src/CoreBundle/Resources/public/javascript/tableEditor.js:20]: Line 20 should be const if `tableEditorBeforeSaveEvent` is not reassigned.

It's not immediately relevant to this PR, but should those variable declarations in lines 17-18 (et. al. the whole file has `var` declarations) also use the newer variable declaration keywords (const or let depending on reassignment; it seems these are also const, I'm not even sure that these are ever read (lines 17-19 that is).)?

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 2]
```diff
@@ -446,8 +446,12 @@ function ReloadMainPage() {
  * tableEditor: save edit table via ajax
  */
 function SaveViaAjaxCustomCallback(customCallbackFunction, closeAfterSave) {
-    if (customCallbackFunction == 'undefined') customCallbackFunction = SaveViaAjaxCallback;
+    if ('undefined' === customCallbackFunction) {
```
Comment (finn-matti) [inline src/CoreBundle/Resources/public/javascript/tableEditor.js:449]: Yoda conditions in js?

File: src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig [hunk 3]
```diff
@@ -59,6 +58,11 @@
             CKEDITOR.on('instanceReady', function () {
                 if (CKEDITOR.lang.de) CKEDITOR.lang.de.common.browseServer = 'CMS Quelle wählen';
                 else if (CKEDITOR.lang.en) CKEDITOR.lang.en.common.browseServer = 'Select CMS Source';
+
+                document.addEventListener('tableEditorBeforeSaveEvent', function(e) {
+                    let data = CKEDITOR.instances['{{ sEditorName }}'].getData();
```
Comment (finn-matti) [inline src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig:63]: This is probably a const?

**Review by finn-matti — CHANGES_REQUESTED** [review]

File: src/CoreBundle/Resources/public/javascript/tableEditor.js (hunk 1 above)
Comment (pixeljunkie) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js]: To not break any code, I will not do refactoring in this release. The whole backend JS code is very legacy and needs some love. But i will use const here for the added lines.

File: src/CoreBundle/Resources/public/javascript/tableEditor.js (hunk 2 above)
Comment (pixeljunkie) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js]: changed

File: src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig (hunk 3 above)
Comment (pixeljunkie) [reply src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig]: removed this and inlined the code instead

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 4]
```diff
@@ -446,8 +446,12 @@ function ReloadMainPage() {
  * tableEditor: save edit table via ajax
  */
 function SaveViaAjaxCustomCallback(customCallbackFunction, closeAfterSave) {
-    if (customCallbackFunction == 'undefined') customCallbackFunction = SaveViaAjaxCallback;
+    if (customCallbackFunction === 'undefined') {
+        customCallbackFunction = SaveViaAjaxCallback;
+    }
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js]: done

File: src/CoreBundle/Resources/public/javascript/tableEditor.js (hunk 1 above)
Comment (finn-matti) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js]: Understood.

File: src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig [hunk 5]
```diff
@@ -7,10 +7,9 @@
     </textarea>
 
     <script type="text/javascript">
-
         var editorStyleSets = '';
         editorStyleSets = {{ aEditorSettings.stylesSet|raw }};
-        if (false === editorStyleSets in CKEDITOR.stylesSet.registered) {
+        if (editorStyleSets in CKEDITOR.stylesSet.registered === false) {
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Resources/views/snippets-cms/TCMSFieldWYSIWYG/cKEditor/editor.html.twig:12]: coderabbit seams right. It needs to be enclosed with () before checking the boolean value

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 6]
```diff
@@ -17,7 +17,11 @@ var _trans_icon = '';
 var _trans_link = '';
 var _trans_password = '';
 var _trans_drag = '';
-
+const tableEditorBeforeSaveEvent = new CustomEvent('tableEditorBeforeSaveEvent', {
+    detail: {
+        description: 'is dispatched in ExecutePostCommand and SaveViaAjax and Save'
+    }
```
Comment (michaelcmichaelis) [inline src/CoreBundle/Resources/public/javascript/tableEditor.js:23]: Usually, the property "details" is used for the custom parameters assigned per event instance, but this is a classic comment (may be use `// ...` ahead this instantiation instead)
```suggestion
```

**Review by michaelcmichaelis — CHANGES_REQUESTED** [review]

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 7]
```diff
@@ -17,7 +17,11 @@ var _trans_icon = '';
 var _trans_link = '';
 var _trans_password = '';
 var _trans_drag = '';
-
+const tableEditorBeforeSaveEvent = new CustomEvent('tableEditorBeforeSaveEvent', {
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js:20]: it is not an CKEditor event. CKEditor uses the event, but the Markdown editor is triggered by this event, too. I would use the original name if their is no reason egainst it.

File: src/CoreBundle/Resources/public/javascript/tableEditor.js [hunk 8]
```diff
@@ -17,7 +17,11 @@ var _trans_icon = '';
...
+    detail: {
+        description: 'is dispatched in ExecutePostCommand and SaveViaAjax' 
+    }
+});
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Resources/public/javascript/tableEditor.js]: done

**Pull request merged**